### PR TITLE
Merge upstream

### LINKED ISSUE #16: Binary format is out of date
The decoder in this branch is out of date with the bulk-memory-operations branch.

In particular, for data segments:

This repo:
```
let segment dat s =
  let index = at var s in
  let offset = const s in
  let init = dat s in
  {index; offset; init}                                                                                                                
```

Bulk memory:
```
let data s =
  match vu32 s with
  | 0x00l ->
    let dmode = at active_zero s in
    let dinit = string s in
    {dinit; dmode}
  | 0x01l ->
    let dmode = at passive s in
    let dinit = string s in
    {dinit; dmode}                                                                                                                     
  | 0x02l ->
    let dmode = at active s in
    let dinit = string s in
    {dinit; dmode}
  | _ -> error s (pos s - 1) "invalid data segment kind"
```

I.e. it appears this repo is interpreting the reserved 0 byte as `i32` index, when it should be a flags bitfield, with bit 6 indicating the presence of a memory segment index.

### PR DESCRIPTION
**Pull request by rossberg**
Merge main spec now that SIMD has landed. Required extending SIMD load/store instructions with optional memory index in both spec and interpreter.

Also fixes #16.

Edit: It's annoying that GitHub provides no way to view the merge changes alone. You'll have to check all places in spec and interpreter that deal with SIMD load/store instructions.

### CODE REVIEW

File: document/core/appendix/algorithm.rst [hunk 1]
```diff
@@ -25,10 +25,10 @@ Types are representable as an enumeration.
 
 .. code-block:: pseudo
 
-   type val_type = I32 | I64 | F32 | F64 | Funcref | Externref
+   type val_type = I32 | I64 | F32 | F64 | V128 | Funcref | Externref
 
    func is_num(t : val_type | Unknown) : bool =
```
Comment (ngzhian) [inline document/core/appendix/algorithm.rst:30]: not related to this PR, do we want to update this algorithm to have `is_vec`?

**Review by ngzhian — APPROVED** [review]
LGTM! Noticed something in the appendix that we might have missed for SIMD.

File: document/core/appendix/algorithm.rst (hunk 1 above)
Comment (rossberg) [reply document/core/appendix/algorithm.rst:30]: Yes! In fact, `is_num` is incorrect as is, and once fixed, `is_vec` is needed in the case for `select`. Probably my oversight, but would you be interested in creating a PR against the main repo?

Comment (ngzhian) [reply document/core/appendix/algorithm.rst:30]: Done, ptal https://github.com/WebAssembly/spec/pull/1399

**Pull request merged**